### PR TITLE
make update notifications non-blocking; add config option

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
stores next version in a cache which is updated by an async process

### CODE REVIEW

File: src/update_notifier.rs [hunk 1]
```diff
@@ -3,117 +3,231 @@
...
+#[derive(Deserialize, Serialize, Debug, PartialEq)]
+pub struct WapmUpdate {
+    /// The data related to the last check on the Github Registry
+    pub last_check: Option<WapmLastCheck>,
+    /// The time when wapm last trigger the notification
+    pub last_notified: Option<DateTime<Utc>>,
+}
+
+impl WapmUpdate {
+    fn default() -> Self {
```
Comment (MarkMcCaskey) [inline src/update_notifier.rs]: ⛳️ instead add `derive(Default)` you can still call `WapmUpdate::default()` and it'll do exactly the same thing

File: src/update_notifier.rs [hunk 2]
```diff
@@ -3,117 +3,231 @@
...
+    }
+    fn load() -> Result<Self, String> {
+        let path = get_wapm_update_file_path();
+        let json_file = File::open(path).map_err(|err| err.to_string())?;
+        let wasm_update: WapmUpdate =
+            serde_json::from_reader(json_file).map_err(|err| err.to_string())?;
+        Ok(wasm_update)
+    }
+    fn load_or_default() -> Self {
+        Self::load().unwrap_or_else(|_e| Self::default())
```
Comment (MarkMcCaskey) [inline src/update_notifier.rs]: ⛳️  if you derive default you can just do `Self::load().unwrap_or_default() `

File: src/update_notifier.rs [hunk 3]
```diff
@@ -3,117 +3,231 @@
...
+            None => Ok(()),
+            Some(last_check) => {
+                let now = Utc::now();
+                if let Some(last_notified) = self.last_notified {
+                    let time_to_check: time::Duration = time::Duration::from_std(
+                        std::time::Duration::from_secs(WAPM_NOTIFICATION_WINDOW),
+                    )
+                    .unwrap();
+                    if now - last_notified < time_to_check {
+                        return Ok(());
```
Comment (MarkMcCaskey) [inline src/update_notifier.rs]: this is duplicating logic from `should_trigger_check`, we can just call it here

File: src/update_notifier.rs [hunk 4]
```diff
@@ -3,117 +3,231 @@
...
-        let mut response: Response = client
-            .get(GITHUB_RELEASE_PAGE)
-            .header(ACCEPT, HeaderValue::from_static("application/json"))
-            .send()
-            .ok()?;
-        let response_content: VersionResponse = response.json().ok()?;
-        response_content.tag_name
-    };
+        Err(e) => {
+            println!("Background check failed: {}", e);
```
Comment (MarkMcCaskey) [inline src/update_notifier.rs]: ⛳️ use `error!` if we want to warn the user

File: src/update_notifier.rs [hunk 5]
```diff
@@ -3,117 +3,231 @@
...
+        }
+    }
+    try_unlock_background_process()
+}
 
-    if version_tag.is_empty() {
-        return None;
+pub fn get_latest_tag() -> Result<String, String> {
+    let builder = Client::builder();
+    println!("get_latest_tag 0");
```
Comment (MarkMcCaskey) [inline src/update_notifier.rs]: debug statements?

File: src/update_notifier.rs [hunk 6]
```diff
@@ -3,117 +3,231 @@
...
+    let client = match proxy::maybe_set_up_proxy() {
+        Ok(Some(proxy)) => builder.proxy(proxy),
+        Ok(None) => builder, //continue without proxy
+        Err(e) => return Err(e.to_string()),
     }
+    .redirect(RedirectPolicy::limited(10))
+    .build()
+    .map_err(|err| err.to_string())?;
+    println!("get_latest_tag 1");
```
Comment (MarkMcCaskey) [inline src/update_notifier.rs]: here too

File: src/update_notifier.rs [hunk 7]
```diff
@@ -3,117 +3,231 @@
...
+
+    let mut response: Response = client
+        .get(GITHUB_RELEASE_PAGE)
+        .header(ACCEPT, HeaderValue::from_static("application/json"))
+        .send()
+        .map_err(|err| err.to_string())?;
+
+    println!("get_latest_tag 2");
+    let response_content: VersionResponse = response.json().map_err(|err| err.to_string())?;
+    println!("get_latest_tag 3");
```
Comment (MarkMcCaskey) [inline src/update_notifier.rs]: and here

File: src/update_notifier.rs (hunk 3 above)
Comment (syrusakbary) [reply src/update_notifier.rs]: I think it was good to have the two concepts separate, as the windows for each might be different:
* When do we schedule a new check (each 24 hours)
* When we show a notification (each 2 hours)

File: src/update_notifier.rs (hunk 5 above)
Comment (syrusakbary) [reply src/update_notifier.rs]: Ups, thanks!

File: src/bin/wapm.rs [hunk 8]
```diff
@@ -133,18 +145,26 @@ fn main() {
...
         eprintln!("Error: {}", e);
-        std::process::exit(-1);
+        exit_code = -1;
     }
+
+    if cfg!(feature = "update-notifications") && maybe_show_update_notification {
+        update_notifier::check_sync();
+    }
+
+    std::process::exit(exit_code);
```
[the remark this answers is not in the document]
Comment (syrusakbary) [reply src/bin/wapm.rs]: Gotcha! Fixed

**Pull request merged**